File: examples/javascript/hello/hello.js
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { FrameAssert, Game, GameHarness, GameSpec, toPixelInt } from "transmute";

const __dirname = dirname(fileURLToPath(import.meta.url));
const spec = GameSpec.parse(readFileSync(join(__dirname, "gamespec.properties"), "utf8"));

class HelloGame extends Game {
  constructor(spec) {
    super(spec.width, spec.height);
    this.spec = spec;
  }
  render() {
    this.context.fillRect(0, 0, this.width, this.height, this.spec.clearColor);
    this.context.fillRect(20, 20, 40, 40, toPixelInt(220, 180, 60));
  }
}

const headless = process.argv.includes("--headless");
if (headless) {
  const h = new GameHarness(() => new HelloGame(spec));
  h.step(1);
  FrameAssert.assertPixel(h.renderer(), 0, 0, spec.clearColor);
  console.log(`hello headless ok hash=0x${FrameAssert.hash(h.renderer()).toString(16)}`);
} else {
  console.log("Use --headless in JS v1");
  process.exit(1);
}

File: examples/python/hello/hello.py
"""Minimal Python hello example."""

from __future__ import annotations

import sys
from pathlib import Path

from transmute import FrameAssert, Game, GameHarness, GameSpec
from transmute.world import to_pixel_int


class HelloGame(Game):
    def __init__(self, spec: GameSpec) -> None:
        super().__init__(spec.width, spec.height)
        self.spec = spec
        self.clear = spec.clear_color

    def render(self) -> None:
        self.context.fill_rect(0, 0, self.width, self.height, self.clear)
        self.context.fill_rect(20, 20, 40, 40, to_pixel_int(220, 180, 60))


def main(argv: list[str] | None = None) -> None:
    argv = argv if argv is not None else sys.argv[1:]
    root = Path(__file__).resolve().parent
    spec = GameSpec.load(root / "gamespec.properties")
    headless = "--headless" in argv

    if headless:
        with GameHarness(lambda: HelloGame(spec)) as h:
            h.step(1)
            FrameAssert.assert_pixel(h.renderer(), 0, 0, spec.clear_color)
            print(f"hello headless ok hash=0x{FrameAssert.hash(h.renderer()):x}")
        return

    print("Windowed mode not implemented in Python v1; use --headless")
    with GameHarness(lambda: HelloGame(spec)) as h:
        h.step(1)
        print("ok")


if __name__ == "__main__":
    main()
